pypeit/spectrographs/keck_lris.py

```diff
@@ -490,6 +498,136 @@ def subheader_for_spec(self, row_fitstbl, raw_header, extra_header_cards=None,
+
+        Returns:
+            :obj:`tuple`: Three `numpy.ndarray`_ and a :class:`~pypeit.spectrographs.slitmask.SlitMask`.
+            Two arrays are the predictions of the slit edges from the slitmask design and
+            one contains the indices to order the slits from left to right in the PypeIt orientation
+
+        """
+        # Re-initiate slitmask
+        if filename is not None:
+            self.get_slitmask(filename)
+        else:
+            msgs.error('The name of a science file should be provided')
+
+        if self.slitmask is None:
+            msgs.error('Unable to read slitmask design info. Provide a file.')
+
+        platescale = self.get_detector_par(det=1)['platescale']
+
+
+        hdu = fits.open(filename)
+        binning = self.get_meta_value(self.get_headarr(hdu), 'binning')
+        _, bin_spat = parse.parse_binning(binning)
+
+        # Slit center
+        slit_coords = SkyCoord(ra=self.slitmask.onsky[:,0], 
+                               dec=self.slitmask.onsky[:,1], unit='deg')
+        mask_coord = SkyCoord(ra=self.slitmask.mask_radec[0],
+                              dec=self.slitmask.mask_radec[1], unit='deg')
+
+        # build an array of values containing the bottom (right) edge of the slits
+        # starting edge
+        left_edges = []
+        #for islit in x_order:
+        for islit in range(self.slitmask.nslits):
+            sep = mask_coord.separation(slit_coords[islit])
+            PA = mask_coord.position_angle(slit_coords[islit])
+            #
+            alpha = sep.to('arcsec') * np.cos(PA-self.slitmask.posx_pa*units.deg)
+            #delta = sep.to('arcsec') * np.sin(PA-self.slitmask.posx_pa*units.deg)
+            dx_pix = (alpha.value-self.slitmask.onsky[islit,2]/2.) / (platescale*bin_spat)
+            # target is the slit number
+            left_edges.append(np.round(dx_pix))
+        left_edges = np.array(left_edges, dtype=int)
+
+        # Build up the right edges
+        right_edges = left_edges + np.round(
+            self.slitmask.onsky[:,2]/(platescale*bin_spat)).astype(int)
+
+        # Center of slit
+        centers = (left_edges + right_edges)/2.
+
+        # Trim down by detector
+        # TODO -- Deal with Mark4
+        max_spat = 2048//bin_spat
+        if ccdnum == 1:
+            if self.name == 'keck_lris_red':
+                good = centers < 0.
+                xstart = max_spat + 160//bin_spat  # The 160 is for the chip gap
+            elif self.name == 'keck_lris_blue':
+                good = centers < 0.
+                xstart = max_spat + 30//bin_spat  
+            else:
+                msgs.error(f'Not ready to use slitmasks for {self.name}.  Develop it!')
+        else:
+            if self.name in ['keck_lris_red', 'keck_lris_blue']:
+                good = centers >= 0.
+                xstart = -48//bin_spat
+            else:             
+                msgs.error(f'Not ready to use slitmasks for {self.name}.  Develop it!')
```

Comment:
It would be better to put the code in the relevant subclass instead of testing for the spectrograph name.

Reply:
i get your thinking but find it a lot easier to do it this way
and easier to keep up.